// openapi-mermaid/generators/base-generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseGenerator = void 0;
var fs_1 = require("fs");
var handlebars_1 = require("handlebars");
var BaseGenerator = /** @class */ (function () {
    function BaseGenerator(generatorOptions, templateFilePath) {
        this.generatorOptions = generatorOptions;
        this.templateFilePath = templateFilePath;
        this.emptyArrayRegex = /, ]/g;
        if (templateFilePath) {
            var templateSource = (0, fs_1.readFileSync)(templateFilePath, { encoding: 'utf8' });
            this.template = (0, handlebars_1.compile)(templateSource);
        }
    }
    BaseGenerator.prototype.generateFile = function (outputFilePath, context) {
        var _a;
        if (this.template) {
            try {
                var content = this.template(context).replace(this.emptyArrayRegex, ']');
                (0, fs_1.writeFileSync)(outputFilePath, content, { encoding: 'utf8' });
                return content;
            }
            catch (err) {
                this.handleTemplateError(context, err);
            }
        }
        else {
            (_a = this.generatorOptions.logger) === null || _a === void 0 ? void 0 : _a.warn("Template for ".concat(this.GeneratorName, " has not been specified"));
        }
        return null;
    };
    BaseGenerator.prototype.handleTemplateError = function (context, err) {
        var _a, _b, _c, _d, _e, _f;
        (_a = this.generatorOptions.logger) === null || _a === void 0 ? void 0 : _a.error("Error executing template: ".concat((_b = this.templateFilePath) !== null && _b !== void 0 ? _b : 'undefined', "."));
        (_c = this.generatorOptions.logger) === null || _c === void 0 ? void 0 : _c.error("This is likely an issue with the template.");
        (_d = this.generatorOptions.logger) === null || _d === void 0 ? void 0 : _d.error("Data: ".concat(JSON.stringify(context)));
        (_e = this.generatorOptions.logger) === null || _e === void 0 ? void 0 : _e.error("Goto: https://github.com/ikemtz/openapi-ts-generator to report an issue if necessary.");
        (_f = this.generatorOptions.logger) === null || _f === void 0 ? void 0 : _f.error(err);
        throw err;
    };
    return BaseGenerator;
}());
exports.BaseGenerator = BaseGenerator;
