// openapi-mermaid/openapidoc-converter.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenApiDocConverter = void 0;
var generator_options_1 = require("./models/generator-options");
var schema_info_1 = require("./models/schema-info");
var OpenApiDocConverter = /** @class */ (function () {
    function OpenApiDocConverter(options, apiDocument) {
        this.options = options;
        this.apiDocument = apiDocument;
        this.endAlphaNumRegex = /[A-z0-9]*$/s;
        this.startNumberregex = /^\d*/;
    }
    OpenApiDocConverter.prototype.convertDocument = function () {
        var entities = this.convertEntities();
        var paths = [];
        return { entities: entities, paths: paths };
    };
    OpenApiDocConverter.prototype.convertEntities = function () {
        var _a, _b, _c;
        var entities = [];
        for (var schemaName in (_a = this.apiDocument.components) === null || _a === void 0 ? void 0 : _a.schemas) {
            if ((_b = this.apiDocument.components) === null || _b === void 0 ? void 0 : _b.schemas[schemaName]) {
                var schemaWrapperInfo = new schema_info_1.SchemaWrapperInfo((_c = this.apiDocument.components) === null || _c === void 0 ? void 0 : _c.schemas[schemaName]);
                if (schemaWrapperInfo.componentSchemaObject.enum) {
                    this.buildSchemaWrapperInfoForEnum(schemaWrapperInfo);
                }
                else {
                    this.buildSchemaWrapperInfo(schemaWrapperInfo);
                }
                schemaWrapperInfo.updateReferenceProperties(this.options);
                var entity = {
                    isEnum: schemaWrapperInfo.isEnum,
                    enumValues: schemaWrapperInfo.enumValues.map(function (t) {
                        return typeof t === 'string' || t instanceof String ? t : __assign(__assign({}, t), { key: t.key || 0 });
                    }),
                    name: schemaName,
                    referenceProperties: schemaWrapperInfo.referenceProperties,
                    valueProperties: schemaWrapperInfo.valueProperties.filter(this.options.valuePropertyTypeFilterCallBack || generator_options_1.defaultFilter),
                    importTypes: this.getImportTypes(schemaName, schemaWrapperInfo),
                };
                entities.push(entity);
            }
        }
        return entities.filter(this.options.typeFilterCallBack || generator_options_1.defaultFilter);
    };
    OpenApiDocConverter.prototype.buildSchemaWrapperInfoForEnum = function (schemaWrapperInfo) {
        var _a;
        var _this = this;
        schemaWrapperInfo.isEnum = true;
        (_a = schemaWrapperInfo.enumValues).push.apply(_a, (schemaWrapperInfo.componentSchemaObject.enum || []).map(function (x) {
            var _a, _b;
            var key = (_a = _this.startNumberregex.exec(x)) === null || _a === void 0 ? void 0 : _a.at(0);
            var name = ((_b = _this.endAlphaNumRegex.exec(x)) === null || _b === void 0 ? void 0 : _b.at(0)) || '';
            return {
                name: name,
                key: key ? +key : 0,
            };
        }));
    };
    OpenApiDocConverter.prototype.buildSchemaWrapperInfo = function (schemaWrapperInfo) {
        for (var propertyName in schemaWrapperInfo.componentSchemaObject.properties) {
            if ((schemaWrapperInfo.propertySchemaObject = schemaWrapperInfo.componentSchemaObject.properties[propertyName]).type && // NOSONAR
                schemaWrapperInfo.propertySchemaObject.type !== 'array') {
                schemaWrapperInfo.valueProperties.push(this.convertSchemaObjectToPropertyType(propertyName, schemaWrapperInfo));
            }
            else if ((schemaWrapperInfo.propertyReferenceObject = schemaWrapperInfo.componentSchemaObject.properties[propertyName]) //NOSONAR
                .$ref) {
                schemaWrapperInfo.referenceProperties.push(this.convertReferenceObjectToPropertyType(propertyName, schemaWrapperInfo));
            }
            else if (schemaWrapperInfo.propertySchemaObject.type === 'array' && schemaWrapperInfo.propertySchemaObject.items) {
                this.convertArray(propertyName, schemaWrapperInfo);
            }
        }
    };
    OpenApiDocConverter.prototype.convertArray = function (propertyName, schemaWrapperInfo) {
        var arraySchemaObject = schemaWrapperInfo.propertySchemaObject.items;
        if (arraySchemaObject.type) {
            schemaWrapperInfo.valueProperties.push(this.convertArrayObjectToValuePropertyType(propertyName, schemaWrapperInfo));
        }
        else {
            schemaWrapperInfo.propertyReferenceObject = schemaWrapperInfo.propertySchemaObject.items;
            schemaWrapperInfo.referenceProperties.push(this.convertArrayObjectToReferencePropertyType(propertyName, schemaWrapperInfo));
        }
    };
    OpenApiDocConverter.prototype.convertSchemaObjectToPropertyType = function (propertyName, schemaWrapperInfo) {
        var required = this.getIsRequired(propertyName, schemaWrapperInfo);
        return {
            required: required,
            name: propertyName,
            isArray: false,
            type: this.getPropertyType(schemaWrapperInfo),
            maxLength: schemaWrapperInfo.propertySchemaObject.maxLength,
            minLength: schemaWrapperInfo.propertySchemaObject.minLength,
            hasMultipleValidators: +required + +!!schemaWrapperInfo.propertySchemaObject.maxLength + +!!schemaWrapperInfo.propertySchemaObject.minLength > 1,
        };
    };
    OpenApiDocConverter.prototype.convertArrayObjectToValuePropertyType = function (propertyName, schemaWrapperInfo) {
        var required = this.getIsRequired(propertyName, schemaWrapperInfo);
        return {
            required: required,
            name: propertyName,
            isArray: true,
            type: this.getPropertyType(schemaWrapperInfo),
            hasMultipleValidators: false,
        };
    };
    OpenApiDocConverter.prototype.convertArrayObjectToReferencePropertyType = function (propertyName, schemaWrapperInfo) {
        return __assign(__assign({}, this.convertReferenceObjectToPropertyType(propertyName, schemaWrapperInfo)), { isArray: true });
    };
    OpenApiDocConverter.prototype.convertReferenceObjectToPropertyType = function (propertyName, schemaWrapperInfo) {
        return {
            name: propertyName,
            referenceTypeName: this.parseRef(schemaWrapperInfo),
            isArray: false,
            required: (schemaWrapperInfo.componentSchemaObject.required || []).indexOf(propertyName) > -1,
        };
    };
    OpenApiDocConverter.prototype.getPropertyType = function (schemaWrapperInfo) {
        if (schemaWrapperInfo.propertySchemaObject.type === 'array' && schemaWrapperInfo.propertySchemaObject.items) {
            return schemaWrapperInfo.propertySchemaObject.items.type;
        }
        else if (schemaWrapperInfo.propertySchemaObject.type === 'integer' && schemaWrapperInfo.propertySchemaObject.enum) {
            return 'string | number';
        }
        else if (schemaWrapperInfo.propertySchemaObject.type === 'integer') {
            return 'number';
        }
        else if (schemaWrapperInfo.propertySchemaObject.format === 'date' || schemaWrapperInfo.propertySchemaObject.format === 'date-time') {
            return 'date';
        }
        else if (schemaWrapperInfo.propertySchemaObject.format === 'uuid') {
            return 'uuid';
        }
        else if (schemaWrapperInfo.propertySchemaObject.type === 'boolean') {
            return 'bool';
        }
        if (!schemaWrapperInfo.propertySchemaObject.type) {
            throw new Error('Invalid Propety Type');
        }
        return schemaWrapperInfo.propertySchemaObject.type;
    };
    OpenApiDocConverter.prototype.parseRef = function (schemaWrapperInfo) {
        var regexResult;
        var result = null;
        if (schemaWrapperInfo.propertyReferenceObject.$ref &&
            // tslint:disable-next-line: no-conditional-assignment
            (regexResult = this.endAlphaNumRegex.exec(schemaWrapperInfo.propertyReferenceObject.$ref)) // NOSONAR
        ) {
            schemaWrapperInfo.propertyReferenceObject.$ref = regexResult[0];
            result = schemaWrapperInfo.propertyReferenceObject.$ref;
        }
        return result || 'unknown';
    };
    OpenApiDocConverter.prototype.getImportTypes = function (entityName, schemaWrapperInfo) {
        return schemaWrapperInfo.referenceProperties
            .map(function (t) { return t.referenceTypeName; })
            .filter(function (t) { return t !== entityName; })
            .filter(function (value, index, array) { return array.indexOf(value) === index; })
            .map(function (value) { return ({ name: value }); });
    };
    OpenApiDocConverter.prototype.getIsRequired = function (propertyName, schemaWrapperInfo) {
        return (((schemaWrapperInfo.componentSchemaObject.required || []).indexOf(propertyName) > -1 ||
            (schemaWrapperInfo.propertySchemaObject.nullable === undefined ? false : !schemaWrapperInfo.propertySchemaObject.nullable)) &&
            propertyName !== 'id');
    };
    return OpenApiDocConverter;
}());
exports.OpenApiDocConverter = OpenApiDocConverter;
